From 713477d4bb900e1257fc691ecebd52d648c44be2 Mon Sep 17 00:00:00 2001
From: Thiemo van Engelen <tvanengelen@victronenergy.com>
Date: Wed, 20 Sep 2023 09:10:11 +0200
Subject: [PATCH 1/2] Cache the UUID of characteristics

This prevents (multiple) synchronous dbus calls to retrieve the UUID on a
read or write of a characteristic value.

This is usefull for Venus system that are used as CI runner for our BLE
products. It is not used in regular Venus functionality.
Upstream-Status: Inappropriate [Venus specific]
---
 src/bluetooth/qlowenergycontroller_bluezdbus.cpp | 9 +++++----
 src/bluetooth/qlowenergycontroller_bluezdbus_p.h | 1 +
 2 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/src/bluetooth/qlowenergycontroller_bluezdbus.cpp b/src/bluetooth/qlowenergycontroller_bluezdbus.cpp
index a00e4383..863d6467 100644
--- a/src/bluetooth/qlowenergycontroller_bluezdbus.cpp
+++ b/src/bluetooth/qlowenergycontroller_bluezdbus.cpp
@@ -658,6 +658,7 @@ void QLowEnergyControllerPrivateBluezDBus::discoverServiceDetails(
                                             QStringLiteral("org.bluez"), it.key().path(),
                                             QDBusConnection::systemBus());
                 GattCharacteristic dbusCharData;
+                dbusCharData.uuid = QBluetoothUuid(charInterface->uUID());
                 dbusCharData.characteristic = charInterface;
                 dbusData.characteristics.append(dbusCharData);
             } else if (iface == QStringLiteral("org.bluez.GattDescriptor1")) {
@@ -1019,7 +1020,7 @@ void QLowEnergyControllerPrivateBluezDBus::scheduleNextJob()
                             service->characteristicList.value(nextJob.handle);
         bool foundChar = false;
         for (const auto &gattChar : std::as_const(dbusServiceData.characteristics)) {
-            if (charData.uuid != QBluetoothUuid(gattChar.characteristic->uUID()))
+            if (charData.uuid != gattChar.uuid)
                 continue;
 
             QDBusPendingReply<QByteArray> reply = gattChar.characteristic->ReadValue(QVariantMap());
@@ -1048,7 +1049,7 @@ void QLowEnergyControllerPrivateBluezDBus::scheduleNextJob()
                             service->characteristicList.value(nextJob.handle);
         bool foundChar = false;
         for (const auto &gattChar : std::as_const(dbusServiceData.characteristics)) {
-            if (charData.uuid != QBluetoothUuid(gattChar.characteristic->uUID()))
+            if (charData.uuid != gattChar.uuid)
                 continue;
 
             QVariantMap options;
@@ -1090,7 +1091,7 @@ void QLowEnergyControllerPrivateBluezDBus::scheduleNextJob()
         const QBluetoothUuid descUuid = charData.descriptorList[nextJob.handle].uuid;
         bool foundDesc = false;
         for (const auto &gattChar : std::as_const(dbusServiceData.characteristics)) {
-            if (charData.uuid != QBluetoothUuid(gattChar.characteristic->uUID()))
+            if (charData.uuid != gattChar.uuid)
                 continue;
 
             for (const auto &gattDesc : std::as_const(gattChar.descriptors)) {
@@ -1134,7 +1135,7 @@ void QLowEnergyControllerPrivateBluezDBus::scheduleNextJob()
         const QBluetoothUuid descUuid = charData.descriptorList[nextJob.handle].uuid;
         bool foundDesc = false;
         for (const auto &gattChar : std::as_const(dbusServiceData.characteristics)) {
-            if (charData.uuid != QBluetoothUuid(gattChar.characteristic->uUID()))
+            if (charData.uuid != gattChar.uuid)
                 continue;
 
             for (const auto &gattDesc : std::as_const(gattChar.descriptors)) {
diff --git a/src/bluetooth/qlowenergycontroller_bluezdbus_p.h b/src/bluetooth/qlowenergycontroller_bluezdbus_p.h
index afa2e8f3..6c110099 100644
--- a/src/bluetooth/qlowenergycontroller_bluezdbus_p.h
+++ b/src/bluetooth/qlowenergycontroller_bluezdbus_p.h
@@ -132,6 +132,7 @@ private:
 
     struct GattCharacteristic
     {
+        QBluetoothUuid uuid;
         QSharedPointer<OrgBluezGattCharacteristic1Interface> characteristic;
         QSharedPointer<OrgFreedesktopDBusPropertiesInterface> charMonitor;
         QList<QSharedPointer<OrgBluezGattDescriptor1Interface>> descriptors;
-- 
2.40.0.windows.1

